### DISCUSSION TITLE
Detailed edit history tracking and management

### DISCUSSION THREAD

**Discussion by omniraptorr** [toeverything/AFFiNE#6290] [Ideas]
I would like to have more fine grained google docs-style history (gdocs seems to store edits word by word, judging from e.g. the draftback extension https://chromewebstore.google.com/detail/draftback/nnajoiemfpldioamchanognpjmocgkbg). Would really love to see similar rewind/playback and history visualization features in affine.

A related idea is an integrated version control system along the lines of `undo-tree` in various text editors, such as https://github.com/mbbill/undotree or https://www.emacswiki.org/emacs/UndoTree. The links explain it better, but the most basic 'argh' moment that such as system solves for me is undoing a bunch of changes to retrieve an old bit of text, then accidentally inputting another character and losing all the 'future' changes before i had a chance to re-apply them.

**Comment by EYHN** [comment]
Great suggestion, in fact, our data structure currently contains such information, all deleted text is actually still stored there.

But we don't record the delete time, that is, we don't know when the corresponding text was deleted.

Based on that, we may not be able to achieve the "rewind" effect, but it is feasible to show all deleted text.

Tell me what you think

**Reply by omniraptorr** [reply]
hmm, imo being able to place the edits in a sequence is necessary to make sense of them. Not sure how useful just having an unordered list of edits would be. Unless I misunderstood what you wrote?
